tapchannel/aux_sweeper.go

```diff
@@ -360,6 +384,190 @@ func commitRevokeSweepDesc(keyRing *lnwallet.CommitmentKeyRing,
+	assetOutput *cmsg.AssetOutput, chainParams *address.ChainParams,
+	commitTx *wire.MsgTx, courierAddr *url.URL) error {
+
+	// Make a vPacket with the input set from the funding input proof for
+	// this asset ID.
+	assetID := assetOutput.AssetID.Val
+	vPkt, ok := vPktsByAssetID[assetID]
+	if !ok {
+		fundingInputProof, ok := fundingInputProofs[assetID]
+		if !ok {
+			return fmt.Errorf("funding input proof not found for "+
+				"asset ID: %v", assetOutput.AssetID)
+		}
+		vPkt, err := tappsbt.FromProofs(
+			[]*proof.Proof{fundingInputProof}, chainParams,
+		)
+		if err != nil {
+			return fmt.Errorf("unable to create "+
+				"vPacket: %v", err)
+		}
+
+		vPktsByAssetID[assetID] = vPkt
+	}
+
+	// With the input set, we'll now add the extra information for the
+	// output to the packet to complete it.
+	assetProof := assetOutput.Proof.Val
+	inclusionProof := assetProof.InclusionProof
+	vOut := &tappsbt.VOutput{
+		Amount:       assetProof.Asset.Amount,
+		AssetVersion: assetProof.Asset.Version,
+		Type: func() tappsbt.VOutputType {
+			if assetProof.SplitRootProof == nil {
+				return tappsbt.TypeSimple
+			}
+			return tappsbt.TypeSplitRoot
+		}(),
+		Interactive:             true,
+		AnchorOutputIndex:       inclusionProof.OutputIndex,
+		AnchorOutputInternalKey: inclusionProof.InternalKey,
+		//nolint:ll
+		AnchorOutputTapscriptSibling: inclusionProof.CommitmentProof.TapSiblingPreimage,
+		ScriptKey:                    assetProof.Asset.ScriptKey,
+		ProofSuffix:                  &assetProof,
+	}
+
+	// While we're here, we'll also replace the transaction stored in the
+	// proof with the correct one.
+	vOut.ProofSuffix.AnchorTx = *commitTx
+
+	// Finally, we'll make the proper control block to spend the OP_TRUE
+	// funding output.
+	fundingWitness, err := fundingSpendWitness().Unpack()
+	if err != nil {
+		return fmt.Errorf("unable to make funding witness: %w", err)
+	}
+
+	// There is always only a single input, which is the funding output.
+	const inputIndex = 0
+	vOut.Asset.UpdateTxWitness(0, fundingWitness)
+
+	// Finally, we'll set the delivery address to the default courier, so
+	// we publish the proof in the specified Universe.
+	vOut.ProofDeliveryAddress = courierAddr
+
+	vPkt.Outputs = append(vPkt.Outputs, vOut)
+	vPktsByAssetID[assetID] = vPkt
+
+	return nil
+}
+
+// importCommitTx imports the commitment transaction into the wallet. This is
+// called after a force close to ensure that we can properly spend outputs
+// created by the commitment transaction at a later step.
+func (a *AuxSweeper) importCommitTx(req lnwallet.ResolutionReq,
+	commitState *cmsg.Commitment, fundingInfo *cmsg.OpenChannel) error {
+
+	// To start, we'll re-create vPackets for all of the outputs of the
+	// commitment transaction.
+	//
+	// We'll use the fundingInfo proofs to create a vIn for each of them.
+	fundingInputProofs := make(map[asset.ID]*proof.Proof)
+	for _, fundingInput := range fundingInfo.FundedAssets.Val.Outputs {
```

Comment:
Hmm, so with all these loops we basically unroll the result of `CreateAllocations` back into a virtual packet... I wonder if there was an easier way to do this, but honestly couldn't think of a better way off the top of my head.

Reply:
Heh yeh, there _def_ is some overlap there. I _did_ use the allocation creation for spending from the commitment outputs however. 

Was going to convert it, then just had a bit of sunken cost I guess. Not too late to consolidate either way. Would need to refactor a bit though, as we don't have things like the filtered htlc view here.